File: python/tests/test_zippedcsv.py
import json
from pathlib import Path
import unittest
from zipfile import ZipFile
from tempfile import TemporaryDirectory

from zippedcsv import Csv, ZippedCsv


class ZippedCsvTests(unittest.TestCase):
    def test_zippedcsv_add_read_and_metadata_optional(self) -> None:
        with TemporaryDirectory() as tmpdir:
            file_path = Path(tmpdir) / "example.zcsv"

            zcsv = ZippedCsv(str(file_path))
            csv = Csv([["a", "b", "c"], ["1", "2", "3"]])
            zcsv.AddCsv("sample.csv", csv)
            zcsv.Save()
            zcsv.Close()

            with ZipFile(file_path, "r") as archive:
                self.assertNotIn("metadata.json", archive.namelist())

            reopened = ZippedCsv(str(file_path))
            table_names = reopened.GetTableNames()
            self.assertIn("sample.csv", table_names)
            loaded = reopened.GetCsv("sample.csv")
            self.assertEqual(loaded.GetHeader(), ["a", "b", "c"])
            self.assertEqual(loaded.GetRow(0), {"a": "1", "b": "2", "c": "3"})

    def test_zippedcsv_metadata_persist_and_apply(self) -> None:
        with TemporaryDirectory() as tmpdir:
            file_path = Path(tmpdir) / "metadata.zcsv"
            zcsv = ZippedCsv(str(file_path))
            zcsv.SetMetadata(
                {
                    "title": "Metadata test",
                    "description": "desc",
                    "author": "tester",
                    "createdAt": "2026-02-27",
                    "updatedAt": "2026-02-27",
                    "delimiter": ";",
                    "quoteChar": '"',
                    "hasHeader": True,
                }
            )
            zcsv.AddCsv("sample.csv", Csv([["a", "b"], ["1", "2"]]))
            zcsv.Save()
            zcsv.Close()

            with ZipFile(file_path, "r") as archive:
                self.assertIn("metadata.json", archive.namelist())
                metadata = json.loads(archive.read("metadata.json"))
                raw_csv = archive.read("sample.csv").decode("utf-8")
                self.assertEqual(metadata["delimiter"], ";")
                self.assertIn(";", raw_csv)

            reopened = ZippedCsv(str(file_path))
            self.assertEqual(reopened.GetMetadata()["delimiter"], ";")
            self.assertEqual(reopened.GetCsv("sample.csv").GetRow(0), {"a": "1", "b": "2"})

    def test_zippedcsv_invalid_metadata_recovery(self) -> None:
        with TemporaryDirectory() as tmpdir:
            file_path = Path(tmpdir) / "broken.zcsv"

            with ZipFile(file_path, "w") as archive:
                archive.writestr("metadata.json", '{"delimiter": ",", broken-json')
                archive.writestr("sample.csv", "a,b\n1,2\n")

            zcsv = ZippedCsv(str(file_path))
            self.assertEqual(zcsv.GetCsv("sample.csv").GetRow(0), {"a": "1", "b": "2"})
            self.assertEqual(zcsv.GetMetadata()["delimiter"], ",")
            zcsv.Save()

            with ZipFile(file_path, "r") as archive:
                metadata = json.loads(archive.read("metadata.json"))
                self.assertIsInstance(metadata, dict)


if __name__ == "__main__":
    unittest.main()

File: js/tests/test_zippedcsv.ts
import { Csv, ZippedCsv } from "zippedcsv";
import AdmZip from "adm-zip";
import * as fs from "fs";
import * as os from "os";
import * as path from "path";

function tempFile(dir: string, name: string): string {
  return path.join(dir, name);
}

describe("ZippedCsv", () => {
  test("add, save, read without metadata", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = tempFile(tmpDir, "example.zcsv");

    const zcsv = new ZippedCsv(filePath);
    const csv = new Csv([["a", "b", "c"], ["1", "2", "3"]]);
    zcsv.addCsv("sample.csv", csv);
    zcsv.save();
    zcsv.close();

    // metadata.json should not exist
    const zip = new AdmZip(filePath);
    const names = zip.getEntries().map((e) => e.entryName);
    expect(names).not.toContain("metadata.json");

    const reopened = new ZippedCsv(filePath);
    expect(reopened.getTableNames()).toContain("sample.csv");
    const loaded = reopened.getCsv("sample.csv");
    expect(loaded.getHeader()).toEqual(["a", "b", "c"]);
    expect(loaded.getRow(0)).toEqual({ a: "1", b: "2", c: "3" });

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("metadata persists and drives CSV delimiter", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = tempFile(tmpDir, "metadata.zcsv");

    const zcsv = new ZippedCsv(filePath);
    zcsv.setMetadata({
      title: "Metadata test",
      description: "desc",
      author: "tester",
      createdAt: "2026-01-01",
      updatedAt: "2026-01-01",
      delimiter: ";",
      quoteChar: '"',
      hasHeader: true,
    });
    zcsv.addCsv("sample.csv", new Csv([["a", "b"], ["1", "2"]]));
    zcsv.save();
    zcsv.close();

    // Verify raw archive contents
    const zip = new AdmZip(filePath);
    const names = zip.getEntries().map((e) => e.entryName);
    expect(names).toContain("metadata.json");
    const meta = JSON.parse(zip.readAsText("metadata.json"));
    expect(meta.delimiter).toBe(";");
    const rawCsv = zip.readAsText("sample.csv");
    expect(rawCsv).toContain(";");

    // Re-open and verify
    const reopened = new ZippedCsv(filePath);
    expect(reopened.getMetadata().delimiter).toBe(";");
    expect(reopened.getCsv("sample.csv").getRow(0)).toEqual({ a: "1", b: "2" });

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("invalid metadata JSON falls back to defaults", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = tempFile(tmpDir, "broken.zcsv");

    // Craft a broken archive manually
    const zip = new AdmZip();
    zip.addFile("metadata.json", Buffer.from('{"delimiter": ",", broken-json'));
    zip.addFile("sample.csv", Buffer.from("a,b\n1,2\n"));
    zip.writeZip(filePath);

    const zcsv = new ZippedCsv(filePath);
    expect(zcsv.getMetadata().delimiter).toBe(",");
    expect(zcsv.getCsv("sample.csv").getRow(0)).toEqual({ a: "1", b: "2" });

    // Re-saving repairs the metadata
    zcsv.save();
    const repaired = new AdmZip(filePath);
    const reparsed = JSON.parse(repaired.readAsText("metadata.json"));
    expect(typeof reparsed).toBe("object");

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("non-root CSV files are ignored", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = tempFile(tmpDir, "nested.zcsv");

    const zip = new AdmZip();
    zip.addFile("root.csv", Buffer.from("a,b\n1,2\n"));
    zip.addFile("subdir/nested.csv", Buffer.from("x,y\n3,4\n"));
    zip.writeZip(filePath);

    const zcsv = new ZippedCsv(filePath);
    const names = zcsv.getTableNames();
    expect(names).toContain("root.csv");
    expect(names).not.toContain("subdir/nested.csv");

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("getName with and without .csv extension", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = tempFile(tmpDir, "ext.zcsv");

    const zcsv = new ZippedCsv(filePath);
    zcsv.addCsv("table", new Csv([["a"], ["1"]]));
    zcsv.save();

    const reopened = new ZippedCsv(filePath);
    // Both forms should work for retrieval
    expect(reopened.getCsv("table").getRow(0)).toEqual({ a: "1" });
    expect(reopened.getCsv("table.csv").getRow(0)).toEqual({ a: "1" });

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("removeCsv removes the table", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = tempFile(tmpDir, "remove.zcsv");

    const zcsv = new ZippedCsv(filePath);
    zcsv.addCsv("a.csv", new Csv([["x"], ["1"]]));
    zcsv.addCsv("b.csv", new Csv([["y"], ["2"]]));
    zcsv.removeCsv("a.csv");
    zcsv.save();

    const reopened = new ZippedCsv(filePath);
    expect(reopened.getTableNames()).not.toContain("a.csv");
    expect(reopened.getTableNames()).toContain("b.csv");

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("getCsv throws for unknown file", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = tempFile(tmpDir, "empty.zcsv");

    const zcsv = new ZippedCsv(filePath);
    expect(() => zcsv.getCsv("missing.csv")).toThrow();

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("getMetadata returns copy", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = tempFile(tmpDir, "copy.zcsv");

    const zcsv = new ZippedCsv(filePath);
    const meta = zcsv.getMetadata();
    meta.title = "mutated";
    expect(zcsv.getMetadata().title).toBe("");

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("invalid delimiter/quoteChar normalized to defaults", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = tempFile(tmpDir, "invalid.zcsv");

    const zcsv = new ZippedCsv(filePath);
    zcsv.setMetadata({ delimiter: "abc", quoteChar: "" });
    const meta = zcsv.getMetadata();
    expect(meta.delimiter).toBe(",");
    expect(meta.quoteChar).toBe('"');

    fs.rmSync(tmpDir, { recursive: true });
  });

  test("creates parent directories on save", () => {
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "zcsv-"));
    const filePath = path.join(tmpDir, "deep", "nested", "test.zcsv");

    const zcsv = new ZippedCsv(filePath);
    zcsv.save();
    expect(fs.existsSync(filePath)).toBe(true);

    fs.rmSync(tmpDir, { recursive: true });
  });
});
